enclave/src/sidechain/validateer.rs

```diff
@@ -0,0 +1,95 @@
+use crate::{error::Error, onchain::storage::GetOnchainStorage, Result};
+use frame_support::ensure;
+use pallet_teerex_storage::{TeeRexStorage, TeerexStorageKeys};
+use sp_core::H256;
+use sp_runtime::traits::Header as HeaderT;
+use sp_std::prelude::Vec;
+use substratee_node_primitives::Enclave;
+
+pub trait ValidateerFetch {
+	fn current_validateers<Header: HeaderT<Hash = H256>>(
+		&self,
+		latest_header: &Header,
+	) -> Result<Vec<Enclave>>;
+	fn validateer_count<Header: HeaderT<Hash = H256>>(&self, latest_header: &Header)
+		-> Result<u64>;
+}
+
+impl<OnchainStorage: GetOnchainStorage> ValidateerFetch for OnchainStorage {
+	fn current_validateers<Header: HeaderT<Hash = H256>>(
+		&self,
+		header: &Header,
+	) -> Result<Vec<Enclave>> {
```

Comment:
interesting pattern here, to implement a trait for another generic type - that's something I should keep in mind

Reply:
It is not used really often. Admittedly, I am unsure, which problems this pattern resolves or creates. But here it felt suitable.